Yamaha device setup enhancement with unique id based on serial

pssc opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Remove receive/zone confusion for media player instances add more debug to allow fixing of issue around uid and startup


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #111108 
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

frenck commented:
Hi there @pssc 👋 

There is still an CLA issue, which need to be solved. Without that, we won't be able to review or merge this PR.

Could you take a look?

../Frenck

pssc committed:
fix server unavailale at HA startup Fixes #111108

Remove receiver zone confusion for mediaplayer instances
fix uniq id based on serial where avaialble
get serial suppiled by discovery for config entries.

pssc commented:
Hi @frenck thanks for the heads up think I have finally beaten git into submission and sorted out the CLA and squash down the number of commits to boot.

pssc committed:
Merge branch 'dev' into yamaha

pssc committed:
Fix linter errors

pssc committed:
Merge branch 'dev' into yamaha

pssc committed:
Merge branch 'dev' into yamaha

pssc committed:
Merge branch 'dev' into yamaha

pssc committed:
Merge branch 'dev' into yamaha

pssc committed:
Merge branch 'dev' into yamaha

pssc committed:
Merge branch 'dev' into yamaha

pssc committed:
ruff format

pssc committed:
Enhance debug to find setup code path for tests

pssc committed:
Enhance debug to find setup code path for tests

pssc committed:
Fix formatting

joostlek commented on homeassistant/components/yamaha/media_player.py:
```diff
@@ -205,24 +228,38 @@ def __init__(self, name, receiver, source_ignore, source_names, zone_names):
             # to avoid a breaking change.
-            self._attr_unique_id = f"{self.receiver.serial_number}_{self._zone}"
+            # Prefix as MusicCast could have used this
+            self._attr_unique_id = f"{DOMAIN}_{self.zctrl.serial_number}_{self._zone}"
```
If I understand correctly, this changes the unique_id of the entities, we should not change that without a migration. For me its unclear what this is actually achieving right now since the unique id is the same but with the domain added in front

pssc replied:
I agree and for Yamaha MusicCast enabled amps that the right solution, I was worried about this if I fixed up auto discovery but given I have the last generation of amp without it and that's 9 years old, I think the code will be good enough as is without this discovery, or at last deal with this then...

pssc replied:
Hi, def _discovery (retrieve more info(serial)) if we can via discovery even for static host entries  now and def async_setup_platform is changed to cope with the receiver not being there on planform init.

pssc committed:
Revered uid chanages as not needed yet and cuases other issues

pssc committed:
Revert "Fix formatting"

This reverts commit f3324868d25261a1466233eeb804f526a0023ca1.

pssc committed:
Fix formatting

pssc committed:
Refector tests to cope with changes to plaform init to get serial numbers

pssc committed:
Update test patch

pssc committed:
Update test formatting

pssc committed:
Merge branch 'dev' into yamaha

pssc committed:
Merge branch 'dev' into yamaha

joostlek commented:
Please stop merging dev into your branch. It only wastes CI resources because nothing sensible changed

pssc commented:
> I dont understand what this change is doing, can you please explain?

Sure it does three things at its core:-

1. fixes/provides a uid for entries with a `host:` configured based on the serial no, this is only available from the discovery info used by the library.
2. It fixes the integration to work if say the amp is powered of during HA start.
3. It Explicit clears up the code it is dealing with zone controllers in all cases.

pssc commented:
> Please stop merging dev into your branch. It only wastes CI resources because nothing sensible changed

Sorry was trying to figure something out to do with the CI tests I will stop playing ;-)

joostlek commented:
Can we split the 3 things in separate PRs?

pssc commented:
Of Couse given infinite time and yes smaller commits in the future if anything a have learnt from the Yamaha and Squeezebox  work I have been doing  , but I don't have the time to do that an retest each PR, it all it been running for a couple of months on both my systems I had to do 1 to understand 2 and found 3 in the process.

joostlek commented:
Understandable, but there's a lot happening in different parts of the integration, so if we could split it it would be more clear what fixes what

pssc commented:
> Understandable, but there's a lot happening in different parts of the integration, so if we could split it it would be more clear what fixes what 

Can I explain it that would be simpler and require less work and I can do now 1 and 2 are well entangled.

Lines 116:143 Tidy the discovery to use zones only and match items to config to allow a serial no to be found,
it falls back id the match fails to the previous non serial config (so item 3) also dependant now on 1

162:165 That basically if the fix but is now dependant on 1 for the receiver being off etc... (item 2)

everything else is tidy to zones and extra debug etc.

Kane610 commented:
Hello! This PR has got too many unrelated things going at the same time.  All these changes makes it really hard to determine the effect. Make the PRs easier to approach and it will cost less time for everyone and you will have a better throughput as a result.

pssc committed:
remove all fixes revert code to only make clear we deal with zones and improve debuging

Kane610 reviewed (approved):
Much clearer, thanks!

pssc commented:
Thank you

joostlek commented on homeassistant/components/yamaha/media_player.py:
```diff
@@ -138,21 +143,24 @@ async def async_setup_platform(
-    for receiver in receivers:
-        if config_info.zone_ignore and receiver.zone in config_info.zone_ignore:
+    for zctrl in zone_ctrls:
+        _LOGGER.info("Receiver zone: %s", zctrl.zone)
```
Please don't log on info

pssc replied:
What are the acceptable uses of info logging

pssc replied:
but not here anyway

joostlek commented on homeassistant/components/yamaha/media_player.py:
```diff
@@ -205,24 +215,38 @@ def __init__(self, name, receiver, source_ignore, source_names, zone_names):
+                self._attr_unique_id,
+            )
+        else:
+            _LOGGER.info(
```
Idem

pssc replied:
here I do sort of want to let the user know.

pssc replied:
where can I use info or should this be warn?

joostlek replied:
Warning or debug

joostlek commented on homeassistant/components/yamaha/media_player.py:
```diff
@@ -260,10 +284,15 @@ def build_source_list(self):
             if source not in self._source_ignore
         )
 
+    @property
+    def unique_id(self) -> str:
+        """Return the unique ID for this media_player."""
+        return self._attr_unique_id or ""
+
```
This is the default implementation, please revert

pssc replied:
will do with the uid fix pull next
